vault/src/execution.rs

```diff
@@ -237,16 +239,135 @@ impl Request {
     ) -> Result<TransactionMetadata, Error> {
         let fee_rate = self.get_fee_rate(parachain_rpc).await?;
 
+        tracing::debug!("Using fee_rate = {fee_rate}");
+
         let tx = btc_rpc
             .create_transaction(self.btc_address, self.amount as u64, fee_rate, Some(self.hash))
             .await?;
         let recipient = tx.recipient.clone();
         tracing::info!("Sending bitcoin to {}", recipient);
 
         let txid = btc_rpc.send_transaction(tx).await?;
+        self.await_inclusion(parachain_rpc, btc_rpc, num_confirmations, txid)
+            .await
+    }
 
-        loop {
-            let tx_metadata = btc_rpc.wait_for_transaction_metadata(txid, num_confirmations).await?;
+    async fn bump_fee<B: BitcoinCoreApi + Clone>(
+        &self,
+        btc_rpc: &B,
+        old_txid: Txid,
+        fee_rate: u64,
+    ) -> Result<Txid, Error> {
+        let new_tx = btc_rpc
+            .create_bumped_transaction(&old_txid, self.btc_address, fee_rate)
+            .await?;
+        let new_txid = btc_rpc.send_transaction(new_tx).await?;
+        tracing::info!("Bumped fee... {new_txid}");
+        Ok(new_txid)
+    }
+
+    #[tracing::instrument(
+        name = "await_inclusion",
+        skip(self, parachain_rpc, btc_rpc),
+        fields(
+            request_type = ?self.request_type,
+            request_id = ?self.hash,
+        )
+    )]
+    async fn await_inclusion<
+        B: BitcoinCoreApi + Clone,
+        P: OraclePallet + BtcRelayPallet + VaultRegistryPallet + UtilFuncs + Clone + Send + Sync,
+    >(
+        &self,
+        parachain_rpc: &P,
+        btc_rpc: &B,
+        num_confirmations: u32,
+        mut txid: Txid,
+    ) -> Result<TransactionMetadata, Error> {
+        'outer: loop {
+            tracing::info!("Awaiting bitcoin confirmations for {txid}");
+
+            let txid_copy = txid.clone();
+
+            let fee_rate_subscription = parachain_rpc.on_fee_rate_change();
+            let fee_rate_subscription = BroadcastStream::new(fee_rate_subscription);
+            let subscription = fee_rate_subscription
+                .map_err(Into::<Error>::into)
+                .and_then(|x| {
+                    tracing::debug!("Received new inclusion fee estimate {}...", x);
+
+                    let ret: Result<u64, _> = x
+                        .into_inner()
+                        .checked_div(FixedU128::accuracy())
+                        .ok_or(Error::ArithmeticUnderflow)
+                        .and_then(|x| x.try_into().map_err(Into::<Error>::into));
+                    futures::future::ready(ret)
+                })
+                .try_filter_map(|x| async move {
+                    match btc_rpc.fee_rate(txid) {
+                        Ok(current_fee) => {
+                            if x >= current_fee {
```

Comment:
I am wondering if we should make this configurable, at least if the fee rate increases significantly enough the Vault might prefer to eat the penalty for non-inclusion right?

Reply:
> the Vault might prefer to eat the penalty for non-inclusion right?

I don't understand what you mean here, can you rephrase/elaborate?

Reply:
Okay so this might be an absurd example, but my point is that we may not want to always blindly increase the fee.

So in reference to the fee budget taken from the oracle when the redeem request is made, let's assume we have the following:
- sat/byte is 100
- estimated redeem size is 384 bytes
- fee budget is 38400 sat or 0.000384 BTC = ~$10
- redeem value is 0.1 BTC (total is 0.100384 BTC)

There is no technical limit to the fee rate so let us assume this increases to 3000 sat/byte which the Vault has to cover out-of-pocket:
- vault now pays 1152000 sat (0.01152 BTC = ~$275) - maybe less depending on actual size
- total sent is now 0.11152 BTC

Now assume the Vault decided not to bump the transaction fees:
- slashed 100% value in collateral (e.g. DOT)
- punishment fee (10%) of 0.1 BTC also taken = 0.01 BTC (this may also be increased in the future)
- total lost is 0.11 BTC - lower net loss than bumping the transaction fee

Reply:
As @sander2 has just reminded me, once the transaction has been sent the Vault anyway risks losing funds if it doesn't bump to include before expiry.